Add support for `vecN<i32>` and `vecN<u32>` to `dot()` function

francesco-cattoglio opened the pull request:
Work-in-progress pull request to fix https://github.com/gfx-rs/naga/issues/1667

Right now I have only implemented this for glsl-out to get an early review.
By reading the discussion at https://github.com/gpuweb/gpuweb/issues/2231 and the minutes in https://docs.google.com/document/d/1o3BrMEZnE3Yat95uLiY06opF2Hmcj5XGlQmBhTT-_dQ/edit# I understand that:
- HLSL requires no extra changes, the `dot()` function will just work 
- MSL and Spir-V will require changes to the code similar to the ones made for GLSL: if vectors of integers are involved we need to turn the `dot()` call into an arithmetic expression.

kvark commented on src/back/glsl/mod.rs:
```diff
@@ -2532,7 +2532,63 @@ impl<'a, W: Write> Writer<'a, W> {
+
+                                        write!(self.out, "(")?;
+
+                                        self.write_expr(arg, ctx)?;
```
Writing the same expression multiple times is a valid thing to do, but it can end up with a lot of unnecessary repetition.
We need to force the arguments of `dot(integer, integer)` to be "baked".
Currently the baking logic in GLSL is just checking `bake_ref_count()`. Here, it needs to change to something like:
```rust
if ref_count > 1 || is_argument_of_integer_dot_product(expr) {
 // bake
}
```

francesco-cattoglio replied:
I tried to find a way to easily bake the arguments of a dot product, but to no avail. If I understand correctly, the "needs to be baked" information should live somewhere in the `FunctionCtx` that is created by each frontend (just like `named_expressions` are), so that when the time comes, the backend can know that the expression needs to be baked. This is a major change to Naga, so I don't think this is something I can achieve.

OTOH, I was wondering: would it be reasonable to do some "local" baking by creating a scope and using a temporary variable? I recon this is a lot of code that handles a very specific case, but perhaps it could be a good compromise.

kvark replied:
The way I thought this could work is you making a `BitSet` to store indices of expressions (within the currently processed functions) that are integer arguments to `dot()` product. When starting each function, you'd clear that bitset, then walk through all expressions and add the relevant indices to the bitset. And then `is_argument_of_integer_dot_product` is basically a check if `expr.index()` is in this bitset.

francesco-cattoglio replied:
I think I get it now, this would be a change that lives mostly only inside the `write_function` call of a backend. I will see if I can manage something reasonable!

kvark commented on src/back/glsl/mod.rs:
```diff
@@ -1136,12 +1136,44 @@ impl<'a, W: Write> Writer<'a, W> {
+        // Do a first pass on all statements and store the handle of the ones
+        // that will need baking
+        use crate::Expression;
+        let mut baking_set = crate::FastHashSet::default();
```
could we make `baking_set` a part of `self` to re-use the allocation between functions?
This whole block should be moved into a function as well.

kvark reviewed (changes requested):
This is looking good!
Just one more thing :)

kvark commented on src/back/msl/writer.rs:
```diff
@@ -1192,7 +1195,67 @@ impl<W: Write> Writer<W> {
+                                match kind {
+                                    ScalarKind::Float => "dot",
+                                    ScalarKind::Sint | ScalarKind::Uint => {
+                                        // No intrinsic function for integer dot product in MSL,
```
let's move this into a function

kvark commented on src/back/msl/writer.rs:
```diff
@@ -1192,7 +1195,67 @@ impl<W: Write> Writer<W> {
                     Mf::Pow => "pow",
                     // geometry
-                    Mf::Dot => "dot",
+                    Mf::Dot => {
```
should just do `=> match *context.resolve_type(arg) {..}` from here

kvark commented on src/back/msl/writer.rs:
```diff
@@ -1192,7 +1195,67 @@ impl<W: Write> Writer<W> {
+                        use crate::TypeInner;
+                        let inner = context.resolve_type(arg);
+                        match *inner {
+                            TypeInner::Vector { kind, size, .. } => {
```
let's pattern match the kind here, i.e.
```rust
TypeInner::Vector { kind: ScalarKind::Float, .. } => "dot",
TypeInner::Vector { size, .. } => {
  self.write_dot_product(...);
}
```

kvark commented on src/back/msl/writer.rs:
```diff
@@ -1192,7 +1195,67 @@ impl<W: Write> Writer<W> {
+
+                                        self.put_expression(arg1, context, false)?;
+                                        write!(self.out, ".x + ")?;
+
```
should refactor this into a loop:
```rust
for component in COMPONENTS[1..size as usize] {
                                            write!(self.out, " + ")?;

                                            self.put_expression(arg, context, false)?;
                                            write!(self.out, ".{} * ", component)?;

                                            self.put_expression(arg1, context, false)?;
                                            write!(self.out, ".{}", component)?;
}
```

kvark commented on src/back/msl/writer.rs:
```diff
@@ -1192,7 +1195,67 @@ impl<W: Write> Writer<W> {
+
+                                        write!(self.out, "(")?;
+
+                                        self.put_expression(arg, context, false)?;
```
since the backend guarantees `arg` and `arg1` to be baked here, we can and should write this directly:
```rust
        let arg0_name = self.named_expressions[&arg];
        let arg1_name = self.named_expressions[&arg1];
        write!(self.out, "+ {}.{} * {}.{}", arg0_name, component, arg1_name, component)?;
```

kvark commented on src/back/spv/block.rs:
```diff
@@ -552,13 +552,175 @@ impl<'w> BlockContext<'w> {
+                                            &[2],
+                                        ));
+                                        let b_z = self.gen_id();
+                                        block.body.push(Instruction::composite_extract(
```
similarly here, let's move it out into a function and try to write as a loop

kvark commented on src/back/spv/block.rs:
```diff
@@ -552,13 +552,175 @@ impl<'w> BlockContext<'w> {
+                    Mf::Dot => {
+                        use crate::TypeInner;
+                        let inner_ty = self.fun_info[arg].ty.inner_with(&self.ir_module.types);
+                        dbg!(inner_ty);
```
stray `dbg!`

kvark commented on src/back/spv/block.rs:
```diff
@@ -552,13 +552,175 @@ impl<'w> BlockContext<'w> {
-                        arg0_id,
-                        arg1_id,
-                    )),
+                    Mf::Dot => {
```
let's also leave a comment to use `VK_KHR_shader_integer_dot_product` when available

kvark reviewed (changes requested):
very nice!

francesco-cattoglio commented:
I implemented the requested changes. On MSL and GLSL back I think I did a pretty good job so far. When refactoring the sum of components into a loop I incorporated the x as well, and therefore there will be an extra "+" sign at the beginning of the produced expression. Let me know if you like it or not. If not, I will move the product of x outside of the loop.

On spv back, I tried my best to reduce repetition, but perhaps I wrote some "clever code" that is just "confusing and obscure" code for something that could have been done much easier.

kvark reviewed (approved):
Nice!

kvark commented on src/back/spv/block.rs:
```diff
@@ -1120,6 +1141,101 @@ impl<'w> BlockContext<'w> {
+        let sum_3_ids: [Word; 2];
+        let sum_4_ids: [Word; 3];
+        use crate::VectorSize;
+        let sum_ids = match size {
```
I think we can simplify this by starting off `OpConstantNull` of type `result_type_id` and just adding to it in a loop

kvark reviewed (changes requested):
One last thing!

kvark reviewed (approved):
great!
